pyop2/base.py

```diff
@@ -3112,6 +3120,7 @@ def pack(self):
     @validate_type(('sparsity', Sparsity, SparsityTypeError),
                    ('name', str, NameTypeError))
     def __init__(self, sparsity, dtype=None, name=None):
+        DataCarrier.__init__(self)
         self._sparsity = sparsity
         self.lcomm = sparsity.lcomm
         self.rcomm = sparsity.rcomm
```

Comment:
This is a good start, but not all the places in which the data in a Dat or matrix can be modified. In `petsc_base.py` the `.vec`, and `vec_wo` context managers also induce modification of the data. (see `VecAccessMixin`). For matrices you should probably update the version in `assemble` (but also other functions that sound like they modify data, like zero_rows/set_values...).

Additionally, the halo exchange functions (global_to_local_end/local_to_global_end) might change data state as well (the user is allowed to call these).

Finally, the big one, a `par_loop` also updates argument state of all arguments other than `READ` arguments. Probably adapt `ParLoop.update_arg_data_state`.

Reply:
Thank you very much for you advice! I do apologize for my late reply, I have completely forgotten about this pending task! Yes, I initially pushed changes to handle `op2.Global` without considering `op2.Dat`/`op2.Mat` because object versioning was at that time needed for `firedrake.Constant` only.

I have just pushed some changes following your comment.

Regarding `assemble`: did you mean `firedrake.assemble` ? My first guess to catch the various assembly calls performed on `base.Mat` was to update the data version in `petsc_base.Mat.assemble` but I am not sure that all assembly calls go through this route. In particular, unlike `firedrake.Matrix`, `firedrake.ImplicitMatrix` does not seem to rely on `base.Mat`. Instead it uses the PETSc matrix object (PETSc.Mat()), and I am not sure how these two play with each other.

What do you think ?

Reply:
Naive question for ParLoop: Why would we need to adapt `update_arg_data_state` (which is called in `ParLoop.compute`) whereas `local_to_global_end`, which is called after, will take into account the access mode to update the data version ?